### PR TITLE
fix: partial application of condition params

### PR DESCRIPTION
**Pull request by willvedd**
## Description
Allowing for the partial application of condition params if the resolution of the condition is true. This change will enable better efforts to resolve conditions with incomplete context _and_ standardize the behavior between `ListUsers` and `Check`.

Consider the following example:

```
model
  schema 1.1
type user
type document
  relations
    define viewer: [user with condXY]

condition condXY(x: bool, y: bool) {
  x || y
}

--

document:1#viewer@user:jon {condition: condXY}
```

In this example, it isn't necessary to require both parameters of `condXY` because only one is required for a truthy result.

This also fixes an issue where partially-applied condition params behave different in Check than in ListObjects. Continuing with the above example, this is how they'd behave:

```
ListObjects(document#viewer, user:jon, context: {x: true}) --> ["document:1"]
Check(document:1#viewer@user:jon, context: {x: true}) --> error (param 'y' missing)
```

Their results are inconsistent because ListObjects will only emit partial param error only if the evaluation is false, but check previously would emit the error regardless of evaluation outcome.

## References
- [Docs: Conditions in OpenFGA](https://openfga.dev/docs/modeling/conditions)

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: assets/tests/abac_tests.yaml [hunk 1]
```diff
@@ -530,7 +530,59 @@ tests:
...
+              type: document
+              relation: viewer
+            errorCode: 2000
+          - request:
+              user: user:jon
+              type: document
+              relation: viewer
+            context:
+              s: "hello"
+            errorCode: 2000
```
Comment (jon-whit) [inline assets/tests/abac_tests.yaml:545]: This should be `document:1` not an error, yeah?

File: assets/tests/abac_tests.yaml [hunk 2]
```diff
@@ -530,7 +530,59 @@ tests:
...
+            context:
+              s: "not hello"
+            errorCode: 2000
+          - request:
+              user: user:jon
+              type: document
+              relation: viewer
+            context:
+              x: 10
+            errorCode: 2000
```
Comment (jon-whit) [inline assets/tests/abac_tests.yaml:559]: This should be `document:1` not an error

File: assets/tests/abac_tests.yaml (hunk 1 above)
Comment (miparnisari) [reply assets/tests/abac_tests.yaml:545]: I thought this too, but... ListObjects is supposed to return an error only if it cannot fulfil `max_results`. In all these tests, `maxResults = 1000`, but we never expect that many results, so we return an error.

We'd have to change ReverseExpansion quite a bit to ignore errors for objects that it has already sent.

File: internal/graph/check.go [hunk 3]
```diff
@@ -884,18 +880,15 @@ func (c *LocalChecker) checkTTU(parentctx context.Context, req *ResolveCheckRequ
...
+				if len(condEvalResult.MissingParameters) > 0 {
+					errs = multierror.Append(errs, condition.NewEvaluationError(
+						t.GetCondition().GetName(),
+						fmt.Errorf("tuple '%s' is missing context parameters '%v'",
+							tuple.TupleKeyToString(t),
+							condEvalResult.MissingParameters),
+					))
+				}
 				continue
 			}
```
Comment (miparnisari) [inline internal/graph/check.go:893]: Why don't we change this directly within `EvaluateTupleCondition()`? we have this in 4 places already, and in listUsers we're going to copy it again

File: assets/tests/abac_tests.yaml (hunk 1 above)
Comment (willvedd) [reply assets/tests/abac_tests.yaml:545]: Because there are two conditions in the model, only one which has enough context to evaluate true, when the graph expands and evaluates `folder with xcond`, there will be an error. As Maria notes above, this _should_ depend on the number of max results but here we are far from the max results.

File: internal/graph/check.go (hunk 3 above)
Comment (willvedd) [reply internal/graph/check.go:893]: We definitely should do as suggested. I will incorporate that change if we agree to this change in principle.

File: internal/condition/eval/eval.go [hunk 4]
```diff
@@ -25,56 +26,69 @@ func EvaluateTupleCondition(
...
+	}
+
+	if len(conditionResult.MissingParameters) > 0 {
+		err = condition.NewEvaluationError(
+			tupleKey.GetCondition().GetName(),
+			fmt.Errorf("tuple '%s' is missing context parameters '%v'",
+				tuple.TupleKeyToString(tupleKey),
+				conditionResult.MissingParameters),
+		)
 	}
```
Comment (willvedd) [inline internal/condition/eval/eval.go:91]: This is the bulk of the change here.

File: internal/condition/eval/eval.go [hunk 5]
```diff
@@ -25,56 +26,69 @@ func EvaluateTupleCondition(
...
-		metrics.Metrics.ObserveEvaluationCost(conditionResult.Cost)
-
-		span.SetAttributes(attribute.Bool("condition_met", conditionResult.ConditionMet))
-		span.SetAttributes(attribute.String("condition_cost", strconv.FormatUint(conditionResult.Cost, 10)))
-		span.SetAttributes(attribute.StringSlice("condition_missing_params", conditionResult.MissingParameters))
-		return &conditionResult, nil
+
+	if conditionName == "" {
+		return true, nil
+	}
```
Comment (willvedd) [inline internal/condition/eval/eval.go:40]: Only appears like a bigger diff due to the inversion of this if statement.

File: internal/condition/eval/eval_test.go [hunk 6]
```diff
@@ -78,104 +77,116 @@ condition correct_ip(ip: string) {
...
-		tupleKey *openfgav1.TupleKey
-		context  map[string]any
-		result   *condition.EvaluationResult
-	}{
+condition condXY(x: bool, y: bool) {
+	x || y
+}`),
+			context:      map[string]interface{}{},
+			conditionMet: false,
+			expectedErr:  "'condXY' - tuple 'document:1#viewer@user:maria' is missing context parameters '[x y]'",
```
Comment (willvedd) [inline internal/condition/eval/eval_test.go:114]: This test is non-deterministic due to the ordering of the missing params `x` and `y`. If this PR is approved in principle, I will alter the tests to make this deterministic.

**Review by miparnisari — DISMISSED** [review]

File: assets/tests/abac_tests.yaml (hunk 1 above)
Comment (jon-whit) [reply assets/tests/abac_tests.yaml:545]: Created an issue to track this work separately.
https://github.com/openfga/openfga/issues/1511

File: internal/condition/eval/eval.go [hunk 7]
```diff
@@ -25,56 +26,69 @@ func EvaluateTupleCondition(
 	tupleKey *openfgav1.TupleKey,
 	typesys *typesystem.TypeSystem,
 	context *structpb.Struct,
-) (*condition.EvaluationResult, error) {
+) (bool, error) {
```
Comment (miparnisari) [inline internal/condition/eval/eval.go:29]: as discussed offline, should we undo this change?

File: internal/condition/eval/eval.go [hunk 8]
```diff
@@ -25,56 +26,69 @@ func EvaluateTupleCondition(
 	tupleKey *openfgav1.TupleKey,
 	typesys *typesystem.TypeSystem,
 	context *structpb.Struct,
-) (*condition.EvaluationResult, error) {
```
Comment (jon-whit) [inline internal/condition/eval/eval.go:28]: (nit non-blocking): I'm not in favor of the changes around the return type, but I don't want to block these changes.

For context - I think it's misleading to return a bool as that can blindly imply to the caller that the condition was evaluable and resolved to `false`. A nil EvaluationResult implies an error in trying to resolve an evaluation outcome, which is distinctly different than returning a concrete bool with an overloaded error. 

The intent behind having the explicit EvaluationResult was 2-fold: 1) it allows us to grow the response of the evaluation outcome, which would include a residual expression in the future (e.g. we can evaluate a sub-branch of the CEL expression but we return a residual unresolvable expression). And 2) it's the callers responsibility to determine what to do with the outcome of the CEL expression evaluation. It's the `eval` package's responsibility to have a clean API boundary for callers to use the information about the evaluation in a way that makes sense to the caller. The way the code is written in these changes implies something about the callers behavior, and so that feels like a leak of responsibility. 

In general the outcome of a CEL evaluation is a tertiary outcome, not a binary one. A CEL expression can evaluate to a binary value if it is resolvable (e.g. `x && y`, where `x=false` or `x || y`, where `y=true`) or it can result in a residual expression (AST), for example `(x || y) && z`, where x=true and y and z are not provided - residual is the expression `true && z` .

Given we're not really leveraging this distinction in this tertiary state I don't want to block this, but I did want to mention that I think this is the wrong direction for this internal libraries return signature.

File: internal/condition/eval/eval_test.go [hunk 9]
```diff
@@ -78,104 +77,116 @@ condition correct_ip(ip: string) {
...
+    define can_view: [user with condXY]
 
-// TestDefaultCELEvaluationCost is used to ensure we don't decreasee the default evaluation cost
-// of CEL expressions, which would break API compatibility.
-//
-// Critical paths involving ABAC Condition evaluations use the EvaluateTupleCondition function,
-// and so we test that directly to give us higher confidence we're not introducing a compatibility
-// issue.
-func TestDefaultCELEvaluationCost(t *testing.T) {
-	tests := []struct {
```
[the remark this answers is not in the document]
Comment (willvedd) [reply internal/condition/eval/eval_test.go]: Brought back.

File: internal/condition/eval/eval.go (hunk 7 above)
Comment (willvedd) [reply internal/condition/eval/eval.go]: Yes. Please see changes.

File: assets/tests/abac_tests.yaml (hunk 2 above)
Comment (miparnisari) [reply assets/tests/abac_tests.yaml:559]: Tracking via https://github.com/openfga/openfga/issues/1511

File: internal/condition/eval/eval_test.go [hunk 10]
```diff
@@ -78,6 +79,97 @@ condition correct_ip(ip: string) {
...
+  relations
+    define can_view: [user with condXY]
+
+condition condXY(x: bool, y: bool) {
+	x || y
+}`),
+			context:        map[string]interface{}{},
+			conditionMet:   false,
+			expectedErr:    "failed to evaluate relationship condition: 'condXY' - tuple 'document:1#viewer@user:maria' is missing context parameters '[x y]'",
+			expectedAltErr: "failed to evaluate relationship condition: 'condXY' - tuple 'document:1#viewer@user:maria' is missing context parameters '[y x]'",
```
Comment (miparnisari) [inline internal/condition/eval/eval_test.go:117]: nice workaround

**Review by miparnisari — APPROVED** [review]

**Comment by jon-whit** [comment]
Looks good to me. I think we need one more heads up sent over to @aaguiarz and @rhamzeh so that the change of behavior is not a surprise.

**Review by jon-whit — APPROVED** [review]

**Review by miparnisari — CHANGES_REQUESTED** [review]
Let's get alignment from @aaguiarz first before merging.

File: internal/graph/check.go [hunk 11]
```diff
@@ -627,15 +626,6 @@ func (c *LocalChecker) checkDirect(parentctx context.Context, req *ResolveCheckR
 					return nil, err
 				}
 
-				if len(condEvalResult.MissingParameters) > 0 {
```
Comment (sergiught) [inline internal/graph/check.go:630]: If I may be able to pitch in, in my opinion we should take a different approach and ensure that we do not support partial application of conditions params and instead always require all the params of a given condition and error if some are not present. This to me feels like a more straightforward approach that sets better expectations going forward.

**Comment by miparnisari** [comment]
I tried the model + tuples + tests in your description and i'm not seeing a difference with `main`. 

```
# /Users/maria.inesparnisari/Desktop/pr-1505.yml
tests:
  - name: 1
    check:
    - user: user:jon
      object: document:1
      context:
       x: true
      assertions:
        viewer: true
    list_objects:
    - user: user:jon
      type: document
      context:
        x: true
      assertions:
        viewer:
          - document:1
```

On `main`:

```
export FGA_STORE_ID=$(fga store create --api-url 'http://localhost:8080' --name="demo" | jq .store.id --raw-output)

fga model write --api-url 'http://localhost:8080' --format json '{"schema_version":"1.1","type_definitions":[{"type":"user","relations":{},"metadata":null},{"type":"document","relations":{"viewer":{"this":{}}},"metadata":{"relations":{"viewer":{"directly_related_user_types":[{"type":"user","condition":"condXY"}]}}}}],"conditions":{"condXY":{"name":"condXY","expression":"x || y","parameters":{"x":{"type_name":"TYPE_NAME_BOOL"},"y":{"type_name":"TYPE_NAME_BOOL"}}}}}'

fga tuple write user:jon viewer document:1 --condition-name condXY

fga model test  --api-url 'http://localhost:8080' --tests=/Users/maria.inesparnisari/Desktop/pr-1505.yml
```


gives

```
(FAILING) 1: Checks (0/1 passing) | ListObjects (1/1 passing)
ⅹ Check(user=user:jon,relation=viewer,object=document:1, context:&map[x:true]): expected=true, got=N/A, error=Check validation error for POST Check with body {"code":"validation_error","message":"failed to evaluate relationship condition: 'condXY' - context is missing parameters '[y]'"}
 with error code validation_error error message: failed to evaluate relationship condition: 'condXY' - context is missing parameters '[y]'
---
# Test Summary #
Tests 0/1 passing
Checks 0/1 passing
ListObjects 1/1 passing
```

With this PR it gives the same result 🤔 

```
(FAILING) 1: Checks (0/1 passing) | ListObjects (1/1 passing)
ⅹ Check(user=user:jon,relation=viewer,object=document:1, context:&map[x:true]): expected=true, got=N/A, error=Check validation error for POST Check with body {"code":"validation_error","message":"failed to evaluate relationship condition: 'condXY' - context is missing parameters '[y]'"}
 with error code validation_error error message: failed to evaluate relationship condition: 'condXY' - context is missing parameters '[y]'
---
# Test Summary #
Tests 0/1 passing
Checks 0/1 passing
ListObjects 1/1 passing
```

Did I miss something?

**Comment by willvedd** [comment]
Closing-out this PR because the approach is likely going to change. Can always re-open if appropriate.
